### ISSUE TITLE
support for FirstoreDataConverter

### ISSUE THREAD

**Issue by goleary** [nandorojo/swr-firestore#40]
I'm curious if you have thought about adding support for `withConverter` by allowing users of this libary to provide their own type converters?

It might enable usage such as this:
```
const postConverter = {
  toFirestore(post: Post): firebase.firestore.DocumentData {
    return {title: post.title, author: post.author};
  },
  fromFirestore(
    snapshot: firebase.firestore.QueryDocumentSnapshot,
    options: firebase.firestore.SnapshotOptions
  ): Post {
    const data = snapshot.data(options)!;
    if(data.title == null || data.author == null) {
      throw new Error('Data from firestore malformed');
    }
    return new Post(data.title, data.author);
  }
};
...
const {error, data}= useCollection('posts', {converter: postConverter});
```
where data ends up being `Post[]` and `add` only accepts objects of type `Post` (that are then run through the converter)

https://firebase.google.com/docs/reference/js/firebase.firestore.FirestoreDataConverter

**Comment by nandorojo** [comment]
Interesting. Sure, I'm open to adding this. The closest thing so far is the`parseDates` field.

**Comment by devdivdev** [comment]
Yes it will be very interesting, for example encoding / decoding a geo point

**Comment by goleary** [comment]
FYI I've started playing around with this [here](https://github.com/goleary/swr-firestore/commit/f4276742cde134febeec1d504612f25151bf528c).

It seems to fit into the library quite well at least for `useCollection`.

There's more work to be done but I'm happy to submit a PR once I get to spend a bit more time with it and get it to a good place.

**Comment by nandorojo** [comment]
@goleary Thanks, looks good. I added some notes, once those are fixed it should be good to PR.

**Comment by cheunjm** [comment]
@goleary @nandorojo any updates on this PR? if not, I am happy to work on this feature

**Comment by nandorojo** [comment]
You're welcome to!
